docs/contributing/999-rfcs/2024-01-29-tests.md

```diff
@@ -0,0 +1,300 @@
+---
+title: RFC Template
+description: Template for creating a new RFC
+---
+
+# Tests
+
+- **Author(s):**: @tsuf239
+- **Submission Date**: {YYYY-MM-DD}
+- **Stage**: draft
+- **Stage Date**: 2024-01-30
+
+This rfc lists a few features that improve test-writing in wing, and strengthen wing's test framework.
+
+1. [scopes](#1-scopes)
+2. [lifecycle hooks](#2-lifecycle-hooks)
+3. [integration with external libs](#3-integration-with-external-libs)
+4. [mocks](#4-mocks)
+5. [spies](#5-spies)
+6. [resource management optimization](#6-resource-management-optimization)
+
+---
+
+## 1. Scopes
+
+### Design
+
+Currently there is a single test scope:
+
+```
+test {
+
+}
+```
+
+Each test run in separation, and the environment (the collection of resources available in the file) is duplicated per test. Since this is the only way available now, the user cannot re-use resources between tests. The test are running on parallel- one on each cloud.Function and each reports separately to the test runner.
+
+#### Suggested:
+
+We'll add another scope (and have two in total- outer scope, and inner scope) for the following structure:
+//TODO: names
```

Comment:
I'd love to have some help here :) 
@ekeren suggested `test` for outerScope and `check`/`monitor` for innerScope
I feel like it's not clear enough, but I'm not sure :)

Reply:
I think check/monitor is intentionally a different concept, meant for testing live applications. This is RFC seems like it's aiming closer towards unit testing.

Reply:
> I think check/monitor is intentionally a different concept, meant for testing live applications. This is RFC seems like it's aiming closer towards unit testing.

It is different concept but also very very similar to an `inflight test` especially when `inflight test`s do not run in isolation

I feel that in cases where things are very similar there is value in consolidating them into a single concept.

Reply:
I am not sure I agree. The fact that two concepts use similar mechanics doesn't mean they should be the same concept.

`inflight test` is also very similar to just a `cloud.Function` but they are different concepts.

There is also a consideration of design effort. I am not sure we have a reasonable way to model the idea of "an inflight test that doesn't run in isolation" (ideas welcome), some sometimes it's just easier to say "a test always runs in isolation and if you want something that doesn't run in isolation use a monitor".

It's all just words.

Reply:
Here is my original idea for combining test (a wing program that contains both preflight and inflight code) and checks.

https://github.com/winglang/wing/issues/5540#issuecomment-1910665881